base/tuple.jl

```diff
@@ -576,12 +576,13 @@ function _eq(t1::Any32, t2::Any32)
 end
 
 const tuplehash_seed = UInt === UInt64 ? 0x77cfa1eef01bca90 : 0xf01bca90
-hash(::Tuple{}, h::UInt) = h + tuplehash_seed
-hash(t::Tuple, h::UInt) = hash(t[1], hash(tail(t), h))
-function hash(t::Any32, h::UInt)
-    out = h + tuplehash_seed
-    for i = length(t):-1:1
-        out = hash(t[i], out)
+function hash(tup::Tuple, h::UInt)
+    @_terminates_locally_meta
+    f = Fix2(hash, h)
+    tuph = map(f, tup)
+    out = xor(tuplehash_seed, h)
+    for g = tuph
```

Comment:
I'm not sure if `for` loops get unrolled for tuples?

Maybe `out = mapfoldl(g -> 3g, -, tuph, init=out)` instead, since I believe `mapfoldl` is unrolled for tuples (that aren't too long)?

Reply:
The tuple is homogeneous here (`tuph isa Tuple{Vararg{UInt}}`), though, so a `for` loop is optimal. See PR #57091, for example.